### PR TITLE
Integrate YOLACT on mmdetection

### PR DESCRIPTION
**Pull request by chongzhou96**

### CODE REVIEW

File: mmdet/models/anchor_heads/yolact_head.py [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+
+                    w = scale * ar * anchor_base_size
+                    h = scale / ar * anchor_base_size
+                    
+                    # This is for backward compatability with a bug where I made everything square by accident
+                    if self.square_anchor:
+                        h = w
+
+                    # prior_data += [x, y, w, h]
+                    prior_data += [x-0.5*w, y-0.5*h, x+0.5*w, y+0.5*h]
```
Comment (yhcao6) [inline mmdet/models/anchor_heads/yolact_head.py]: The comment seems wrong, should be [x1, y1, x2, y2]?
The coordinates in `make_priors` may not be consistent with other parts like [bbox2delta](https://github.com/open-mmlab/mmdetection/blob/4a4ffd4c8fac40b115bf4485b2fdaac29d89aa75/mmdet/core/bbox/transforms.py#L6) which is called at  https://github.com/open-mmlab/mmdetection/blob/4a4ffd4c8fac40b115bf4485b2fdaac29d89aa75/mmdet/core/anchor/anchor_target.py#L134
```
Code of bbox2delta:
px = (proposals[..., 0] + proposals[..., 2]) * 0.5
py = (proposals[..., 1] + proposals[..., 3]) * 0.5
pw = proposals[..., 2] - proposals[..., 0] + 1.0
ph = proposals[..., 3] - proposals[..., 1] + 1.0
```
In `bbox2delta`, given a anchor, the width should be x2 - x1 + 1.
If using the above coordinate, the width should be x2 - x1.

**Comment by chongzhou96** [comment]
Thanks for your feedback. I'll get into it as soon as possible.

File: mmdet/models/anchor_heads/yolact_head.py [hunk 2]
```diff
@@ -0,0 +1,601 @@
...
+            label_weights = unmap(label_weights, num_total_anchors,
+                                  inside_flags)
+            bbox_targets = unmap(bbox_targets, num_total_anchors, inside_flags)
+            bbox_weights = unmap(bbox_weights, num_total_anchors, inside_flags)
+
+        return (labels, label_weights, bbox_targets, bbox_weights, pos_inds,
+                neg_inds, sampling_result)
+
+
+class YolactAnchorGenerator(AnchorGenerator):
```
Comment (yhcao6) [inline mmdet/models/anchor_heads/yolact_head.py]: The difference between `YolactAnchorGenerator ` and `AnchorGenerator` is minor. Have you tried training with `AnchorGenerator`, will the performance drop?

File: mmdet/models/anchor_heads/yolact_head.py [hunk 3]
```diff
@@ -0,0 +1,601 @@
...
+        self.fp16_enabled = False
+
+        self.anchor_generators = []
+        for anchor_base in self.anchor_base_sizes:
+            self.anchor_generators.append(
+                YolactAnchorGenerator(anchor_base, anchor_scales,
+                                      anchor_ratios))
+
+        self.num_anchors = len(self.anchor_ratios) * len(self.anchor_scales)
+        self._init_layers()
```
Comment (yhcao6) [inline mmdet/models/anchor_heads/yolact_head.py]: Please rearrange above lines following the order of `__init__` function of [`AnchorHead`](https://github.com/open-mmlab/mmdetection/blob/master/mmdet/models/anchor_heads/anchor_head.py)

File: mmdet/models/anchor_heads/yolact_head.py [hunk 4]
```diff
@@ -0,0 +1,601 @@
...
+
+    def init_weights(self):
+        for m in self.head_convs:
+            xavier_init(m.conv, distribution='uniform', bias=0)
+        for m in self.cls_convs:
+            xavier_init(m.conv, distribution='uniform', bias=0)
+        for m in self.box_convs:
+            xavier_init(m.conv, distribution='uniform', bias=0)
+        for m in self.coeff_convs:
+            xavier_init(m.conv, distribution='uniform', bias=0)
```
Comment (yhcao6) [inline mmdet/models/anchor_heads/yolact_head.py]:
```
for module_list in (self.head_convs, self.cls_convs, self.box_convs, self.coeff_convs):
    for m in module_list:
        xavier_init(m.conv, distribution='uniform', bias=0)
```

File: mmdet/models/anchor_heads/yolact_head.py [hunk 5]
```diff
@@ -0,0 +1,601 @@
...
+            sampling=False,
+            unmap_outputs=False,
+            return_sampling_results=return_sampling_results)
+        if return_sampling_results:
+            (labels_list, label_weights_list, bbox_targets_list,
+             bbox_weights_list, num_total_pos, num_total_neg,
+             sampling_results) = cls_reg_targets
+        else:
+            (labels_list, label_weights_list, bbox_targets_list,
+             bbox_weights_list, num_total_pos, num_total_neg) = cls_reg_targets
```
Comment (yhcao6) [inline mmdet/models/anchor_heads/yolact_head.py]: Always return `sampling_results` here.

File: mmdet/models/anchor_heads/yolact_head.py [hunk 6]
```diff
@@ -0,0 +1,601 @@
...
+                      img_metas,
+                      target_means,
+                      target_stds,
+                      cfg,
+                      gt_bboxes_ignore_list=None,
+                      gt_labels_list=None,
+                      label_channels=1,
+                      sampling=True,
+                      unmap_outputs=True,
+                      return_sampling_results=False):
```
Comment (yhcao6) [inline mmdet/models/anchor_heads/yolact_head.py]: It seems the parameter `return_sampling_results` is redundant, we can delete it.

File: mmdet/models/anchor_heads/yolact_head.py [hunk 7]
```diff
@@ -0,0 +1,601 @@
...
+        return dict(loss_cls=losses_cls, loss_bbox=losses_bbox)
+
+    def loss_single(self, cls_score, bbox_pred, labels, label_weights,
+                    bbox_targets, bbox_weights, num_total_samples, cfg):
+        labels = labels.reshape(-1)
+        label_weights = label_weights.reshape(-1)
+        cls_score = cls_score.permute(0, 2, 3,
+                                      1).reshape(-1, self.cls_out_channels)
+        loss_cls_all = F.cross_entropy(
+            cls_score, labels, reduction='none') * label_weights
```
Comment (yhcao6) [inline mmdet/models/anchor_heads/yolact_head.py]: It may be better to make the loss configurable like [here](https://github.com/open-mmlab/mmdetection/blob/7198451afced96091a817a31487ee5038a303aa2/mmdet/models/anchor_heads/anchor_head.py#L43).

File: mmdet/models/mask_heads/yolact_segm_head.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+@HEADS.register_module
+class YolactSegmHead(nn.Module):
+
+    def __init__(
+        self,
+        in_channels=256,
+        num_classes=81,
+        loss_segm=dict(
+            type='CrossEntropyLoss', use_sigmoid=True, loss_weight=1.0),
```
Comment (yhcao6) [inline mmdet/models/mask_heads/yolact_segm_head.py]: Remove last comma.

File: mmdet/models/detectors/yolact.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+        self.segm_head.init_weights()
+        self.mask_head.init_weights()
+
+    def process_gt_single(self, gt_bboxes, gt_labels, gt_masks, device):
+        """Cuda the gt_masks and discard boxes that are smaller than we'd like
+        """
+        gt_masks = torch.from_numpy(gt_masks).to(device)
+        w = gt_bboxes[:, 2] - gt_bboxes[:, 0]
+        h = gt_bboxes[:, 3] - gt_bboxes[:, 1]
+        keep = (w > 4.0) * (h > 4.0)
```
Comment (yhcao6) [inline mmdet/models/detectors/yolact.py]: Make 4.0 configurable. For example add parameters `min_box_width` and `min_box_height`

File: mmdet/core/post_processing/bbox_nms.py [hunk 10]
```diff
@@ -64,3 +64,135 @@ def multiclass_nms(multi_bboxes,
...
+        jaccard overlap: (tensor) Shape: [box_a.size(0), box_b.size(0)]
+    """
+    use_batch = True
+    if box_a.dim() == 2:
+        use_batch = False
+        box_a = box_a[None, ...]
+        box_b = box_b[None, ...]
+
+    inter = intersect(box_a, box_b)
+    area_a = ((box_a[:, :, 2] - box_a[:, :, 0]) *
```
[the remark this answers is not in the document]
Comment (chongzhou96) [reply mmdet/core/post_processing/bbox_nms.py]: Changed to `box_a[:, :, 2] - box_a[:, :, 0] + 1`. After changing, the mask mAP drops 0.1 for both resnet-50 and resnet-101 and box mAP stays the same. But for consistency, I'll keep the `+1`.

File: mmdet/models/anchor_heads/yolact_head.py (hunk 7 above)
Comment (chongzhou96) [reply mmdet/models/anchor_heads/yolact_head.py]: Because we use OHEM, I followed the [code](https://github.com/open-mmlab/mmdetection/blob/7198451afced96091a817a31487ee5038a303aa2/mmdet/models/anchor_heads/ssd_head.py#L113) in SSD. I'll leave it unchanged for now.

File: mmdet/core/post_processing/bbox_nms.py [hunk 11]
```diff
@@ -64,3 +64,135 @@ def multiclass_nms(multi_bboxes,
...
+    iou.triu_(diagonal=1)
+    iou_max, _ = iou.max(dim=1)
+
+    # Now just filter out the ones higher than the threshold
+    keep = (iou_max <= iou_thr)
+
+    # We should also only keep detections over the confidence threshold,
+    # but at the cost of maxing out your detection count for every image,
+    # you can just not do that. Because we have such a minimal amount of
+    # computation per detection (matrix mulitplication only), this increase
```
[the remark this answers is not in the document]
Comment (chongzhou96) [reply mmdet/core/post_processing/bbox_nms.py]: Done. Thanks.

File: mmdet/models/detectors/yolact.py [hunk 12]
```diff
@@ -19,12 +19,27 @@ def __init__(self,
...
         self.mask_head = builder.build_head(mask_head)
+        self._init_weights()
+
+    def _init_weights(self, pretrained=None):
+        self.segm_head.init_weights()
         self.mask_head.init_weights()
 
+    def process_gt_single(self, gt_bboxes, gt_labels, gt_masks, device):
+        """Cuda the gt_masks and discard boxes that are smaller than we'd like
+        """
```
[the remark this answers is not in the document]
Comment (chongzhou96) [reply mmdet/models/detectors/yolact.py]: Done. Thanks.

File: mmdet/models/detectors/yolact.py [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+        x = self.extract_feat(img)
+
+        gt_bboxes, gt_labels, gt_masks = multi_apply(
+            self.process_gt_single,
+            gt_bboxes,
+            gt_labels,
+            gt_masks,
+            device=img.device)
+
+        (cls_score, bbox_pred, coeff_pred) = self.bbox_head(x)
```
[the remark this answers is not in the document]
Comment (chongzhou96) [reply mmdet/models/detectors/yolact.py]: Done. Thanks.

File: mmdet/models/anchor_heads/yolact_head.py (hunk 2 above)
Comment (chongzhou96) [reply mmdet/models/anchor_heads/yolact_head.py]: I removed `YolactAnchorGenerator `

File: mmdet/models/anchor_heads/yolact_head.py (hunk 4 above)
Comment (chongzhou96) [reply mmdet/models/anchor_heads/yolact_head.py]: Done. Thanks.

File: mmdet/models/anchor_heads/yolact_head.py (hunk 6 above)
Comment (chongzhou96) [reply mmdet/models/anchor_heads/yolact_head.py]: Done. Thanks.

File: mmdet/models/mask_heads/yolact_segm_head.py (hunk 8 above)
Comment (chongzhou96) [reply mmdet/models/mask_heads/yolact_segm_head.py]: Done. Thanks.

File: mmdet/models/detectors/yolact.py (hunk 9 above)
Comment (chongzhou96) [reply mmdet/models/detectors/yolact.py]: Done. It's called `min_box_wh` in the `train_cfg`.

File: mmdet/models/mask_heads/yolact_protonet.py [hunk 14]
```diff
@@ -0,0 +1,275 @@
...
+            if isinstance(m, nn.Conv2d):
+                xavier_init(m, distribution='uniform')
+
+    def forward(self, x, coeff_pred, bboxes, img_meta, sampling_results=None):
+        prototypes = self.protonet(x)
+        prototypes = prototypes.permute(0, 2, 3, 1).contiguous()
+
+        num_imgs = x.size(0)
+        # Training state
+        if sampling_results is not None:
```
Comment (yhcao6) [inline mmdet/models/mask_heads/yolact_protonet.py]: We can decide if is training state by `self.training`.

File: mmdet/models/mask_heads/yolact_protonet.py [hunk 15]
```diff
@@ -0,0 +1,275 @@
...
+
+    def get_seg_masks(self, mask_pred, label_pred, img_meta, rescale):
+        ori_shape = img_meta[0]['ori_shape']
+        scale_factor = img_meta[0]['scale_factor']
+        if rescale:
+            img_h, img_w = ori_shape[:2]
+        else:
+            img_h = np.round(ori_shape[0] * scale_factor[1]).astype(np.int32)
+            img_w = np.round(ori_shape[1] * scale_factor[0]).astype(np.int32)
+            scale_factor = 1.0
```
Comment (yhcao6) [inline mmdet/models/mask_heads/yolact_protonet.py]: This line is redundant.

File: mmdet/models/mask_heads/yolact_protonet.py [hunk 16]
```diff
@@ -0,0 +1,275 @@
...
+            cur_img_meta = img_meta[idx]
+            cur_sampling_results = sampling_results[idx]
+
+            pos_assigned_gt_inds = cur_sampling_results.pos_assigned_gt_inds
+            num_pos = pos_assigned_gt_inds.size(0)
+            if num_pos * cur_gt_masks.size(0) == 0:
+                continue
+            if num_pos > self.max_masks_to_train:
+                perm = torch.randperm(num_pos)
+                select = perm[:self.max_masks_to_train]
```
Comment (yhcao6) [inline mmdet/models/mask_heads/yolact_protonet.py]: `select = torch.randperm(num_pos)[self.max_masks_to_train]`. Try to minimize variables used only once.

File: mmdet/models/mask_heads/yolact_protonet.py [hunk 17]
```diff
@@ -0,0 +1,275 @@
...
+            mask_targets = cur_gt_masks[pos_assigned_gt_inds]
+
+            cur_mask_pred = torch.clamp(cur_mask_pred, 0, 1)
+            loss = F.binary_cross_entropy(
+                cur_mask_pred, mask_targets,
+                reduction='none') * self.loss_mask_weight
+
+            h, w = cur_img_meta['img_shape'][:2]
+            gt_bboxes_width = (gt_bboxes_for_reweight[:, 2] -
+                               gt_bboxes_for_reweight[:, 0]) / w
```
Comment (yhcao6) [inline mmdet/models/mask_heads/yolact_protonet.py]:
```
gt_bboxes_width = (gt_bboxes_for_reweight[:, 2] -  gt_bboxes_for_reweight[:, 0] + 1) / w
```

**Comment by lxtGH** [comment]
@chongzhou96 Hi what is performance using mmdet?
I run the r50 model but I found the loss is so big after 2 epoch.

2020-05-28 19:33:45,438 - mmdet - INFO - Epoch [3][1620/1833]   lr: 0.00100, eta: 1 day, 2:25:09, time: 1.183, data_time: 0.067, memory: 21375, loss_cls: 6.7697, loss_bbox: 4.7245, loss_segm: 8.5685, loss_mask: 6.1105, loss: 26.1733
2020-05-28 19:34:00,242 - mmdet - INFO - Epoch [3][1630/1833]   lr: 0.00100, eta: 1 day, 2:26:26, time: 1.481, data_time: 0.074, memory: 21375, loss_cls: 7.3516, loss_bbox: 4.8788, loss_segm: 8.1412, loss_mask: 6.3744, loss: 26.7459
